### PR TITLE
Added the option to use "configbits", see details.

### PR DESCRIPTION
**Pull request by mAPBhlJ**
Many thanks for the awesome library and your contribution on my issue (RSSI), it is of great help to our project (I didn't really know a nice way to implement that myself).
However, I've had another issue, which I was able to resolve myself, and would like to merge it into your project (to have one great repository and not multiple forks with different features).

When sending multiple images after each other, the matrix "resets".
I've asked Senic on how to fix this and they told me they used the last 7 bits of the buffer as "configbits" (see below).

Currently they only have "Onion Skinning" as a possibility, but my code allows any option.
Though, it may to be as user friendly, but it's simple and does the job.

To Senic:

> When displaying a "bar" or "numbers" that you can change by turning, the "fade-in" effect makes it very hard to see the changes. Because every step isn't at 100% brightness immediately, it's hard to scroll through all the steps. (and it looks like a new bar every time, instead of one bar that can be adjusted.
> 
> Can the fade-in effect be disabled?
> 
> I'm using the NuimoJS library by NathanKunicki but I don't believe the fade-in effect (or swiping) has anything to do with the NuimoJS library.
> Does the xcode-swift project allow fade-in to be disabled?
> (I'm using NuimoJS for portability/simplicity reasons)

From Senic:

> Our official SDKs takes care of so called Onion-Skinning Effect. I can explain how it is enabled/disabled. Nuimo has 9x9 LED Matrix that contains 81 LEDs. We need 81 bits to bitmap the whole LED Matrix. Additionally a byte to set the brightness and a byte to set display duration. Therefore totally 13 bytes must be sent to LED characteristics. First 11 bytes represents the bitmap. 11 x 8 bits => 88 bits. In that only 81 bits are used for LEDs and the rest 7 bits are used for other configurations such as enabling Oninon-Skinning effect.  In 11th byte [000X 0001], 5th bit (X) is OnionSkin Bit. So if you don't want to have that effect, then you should keep all the last 7 bits zero.

I'm not sure if we should make a new version out of it, or if we can slide it into the 0.2.0, or put it in a later 0.3.0, but that's up to you I think.

If you need any details, please ask.

### CODE REVIEW

**Comment by nathankellenicki** [comment]
Hi Paul, thanks for the PR! This is a cool feature I didn't know existed, it's great that Senic got back to you about it.

Given that (by their own admission) there may be more features added to the config bits in the future, perhaps it would be more futureproof (and be more user friendly) to pass in an options map rather than an integer? ie. options = {onion-skinning: false}. The bitfield can then be constructed behind-the-scenes, with some sensible defaults if no options are provided.

**Comment by nathankellenicki** [comment]
Ps. A new feature such as this would definitely warrant a bump to 0.3.0. :)

**Comment by mAPBhlJ** [comment]
Hmm, well the sensible default would be to not change the configuration bits, setting them to 0 may not be the best option, since the user may pass a buffer with the config bits changed already (but won't give extra options).

"16" decimal is 0001 0000 binary (onion-skinning).
" 8" decimal is 0000 1000 binary (other flag, not yet implemented by Nuimo).
(Or we could use 0b00010000 and 0b00001000 instead of decimal?)

Fun thing, however, is that you can add up these values.
So we could make constants (or objects?) that you can add up to give a configuration result.
Probably declare something like:

```
//These bitflags can be used in the "setLEDMatrix" function as 4th argument.
//You can add these up by simply putting "config_onion_skinning + config_flicker" as argument, to enable both effects.
const config_onion_skinning_bitflag = 0b00010000;
const config_flicker_bitflag = 0b00001000;
```

So the user could simply do: 
`setLEDMatrix (matrixData, brightness, timeout, config_onion_skinning + config_flicker);`
If he wanted to have onion skinning and flickering effect for example.

I believe that's a simple and efficient way.
There could be a more intuitive way, but I'm more used to C/C++ programming (embedded) and I'm not sure how such a thing is commonly done in JavaScript. (Probably should use an enum or objects that evaluate to these values?)

**Comment by nathankellenicki** [comment]
Yes, I think a more Javascript'y way would be to do something like:

```
device.setLEDMatrix(matrixData, 255, 100, {
    onionSkinning: false,
    configFlicker: true
});
```

As you say, the options would be checked and then added up to make the final int. The default would likely be an empty object ({}), which would of course end with all the bits being 0.

**Comment by larsblumberg** [comment]
Hey, here's Lars from Senic. Here're the extra bits (11th byte) that Nuimo's latest firmware (2.3.2) supports:
5th bit: Enable fading between the previous and the new matrix ("onion skinning")
6th bit: Show built-in matrix or matrix animation. If set it takes the very first byte (of all 13) to specify which matrix to show. Currently we only support 1st byte  = 1. It will show a "waiting" animation.

I'm wondering what you intend with `configFlicker`?

**Comment by mAPBhlJ** [comment]
It was just an example honestly, but was thinking of a configbit that would make the image fade in and out until it reaches the timeout. So that you wouldn't have to send it every second.

But I think it's up to you guys (with feedback from community) to choose which options are interesting, mine was just as an example.

My only problem is that these options aren't documented :p

**Comment by mAPBhlJ** [comment]
We've got two options:

`setLEDMatrix(exampleMatrix, 255, 1000, {onion-skinning: true, builtin-matrix: true});`

or

`setLEDMatrix(exampleMatrix, 255, 1000, config_onion_skinning + config_builtin_matrix);`

With the code:

```
if(options.onionskinning == true){
    buf[10] += 0b00010000;
}
if(options.builtinmatrix == true){
    buf[10] += 0b00100000;
}
```

or

```
(declared somewhere)
const config_onion_skinning = 0b00010000;
const config_builtin_matrix = 0b00100000;

(in function)
if(typeof configbits === "number"){
    buf[10] += configbits;
}
```

respectively.

To me, it doesn't really matter, though I think the second one is simple and quite versatile (any future configbit can be declared anywhere in the code (or just passed as an argument) and it would work fine)

The first one looks a little more fancy,  but the user can't just simply do config_ (autocomplete) to see what options there are. And any new option should be explicitly programmed in, since the boolean options have to be manually converted to the bitflag and put into the configbits.

**Comment by nathankellenicki** [comment]
I see your point about future proofing against future additions, especially if this library falls out of maintenance. (Heaven forbid!)

Perhaps we can have the best of both worlds? Similarly to how the matrixData argument works, we could have type checking to allow both a bitfield (Number) and an options map (Object).

Regarding the consts, could you declare and export them under the Nuimo namespace similar to Nuimo.Swipe.LEFT etc? ie. Nuimo.LEDMatrix.ONION_SKINNING.

Ps. Thanks for the feedback Lars! Great to know about these undocumented features.

**Comment by mAPBhlJ** [comment]
That's actually a pretty neat idea!
The configbits as a number is a good fallback.
And configbits as object for ease of use.

I'll work on it (probably) tomorrow, shouldn't be that much of a problem

**Comment by mAPBhlJ** [comment]
V0.3.0 (Added options for setLEDMatrix) is ready :)

File: src/device.js [hunk 1]
```diff
@@ -233,6 +247,14 @@ class Device extends EventEmitter {
             if(typeof configbits === "number"){//Senic explained that they use config bits
                 buf[10] += configbits;//These are encoded in the unused bits of the buffer
             }
+            if(typeof options === "object"){
+                if(options.onion_skinning == true){
+                    buf[10] += 0b00010000;
```
Comment (nathankellenicki) [inline src/device.js]: Could you use the consts here instead of duplicating the values in the source file?

File: src/device.js [hunk 2]
```diff
@@ -233,6 +247,14 @@ class Device extends EventEmitter {
...
+                }
+                if(options.builtin_matrix == true){
+                    buf[10] += 0b00100000;
```
Comment (nathankellenicki) [inline src/device.js]: Same here. :)

**Comment by nathankellenicki** [comment]
Just a small minor note about duplication of values, but other than that, looks good! :)

**Comment by mAPBhlJ** [comment]
Ah, I see, I developed both options apart from each other.
The consts were for the bitflag adding arguments: "Nuimo.Options.ONION_SKINNING + Nuimo.Options.BUILTIN_MATRIX".

And the Other piece of code was for the javascript way of optional arguments.

But I agree it's more maintainable/logical to have these combined.

---

Fixed in latest edit.

**Comment by nathankellenicki** [comment]
Thanks! Looks great, I've just merged - will push 3.0.0 to npm. :)

**Pull request merged**